crates/nu-command/src/network/url/join.rs

```diff
@@ -0,0 +1,342 @@
+                        }),
+                        "query" => {
+                            if let Some(q) = self.query {
+                                if q != s {
+                                    // if query is present it means that also params_span is setted.
+                                    return Err(ShellError::IncompatibleParameters {
+                                        left_message: format!("Mismatch, query param is: {}", s),
+                                        left_span: value.expect_span(),
+                                        right_message: format!("instead qs from params is: {}", q),
+                                        right_span: self
+                                            .params_span
+                                            .expect("params span should be present"),
+                                    });
+                                }
+                            }
+
+                            Ok(Self {
+                                query: Some(format!("?{}", s)),
+                                query_span: Some(value.expect_span()),
+                                ..self
+                            })
+                        }
+                        "fragment" => Ok(Self {
+                            fragment: Some(if s.starts_with('#') {
+                                s
+                            } else {
+                                format!("#{}", s)
+                            }),
+                            ..self
+                        }),
+                        _ => Ok(self),
+                    }
+                }
+            }
+            _ => Ok(self),
+        }
+    }
+
+    pub fn to_url(&self, span: Span) -> Result<String, ShellError> {
+        let mut user_and_pwd: String = String::from("");
+
+        if let Some(usr) = &self.username {
+            if let Some(pwd) = &self.password {
+                user_and_pwd = format!("{}:{}@", usr, pwd);
+            }
+        }
+
+        let scheme_result = match &self.scheme {
+            Some(s) => Ok(s),
+            None => Err(UrlComponents::generate_shell_error_for_missing_parameter(
+                String::from("scheme"),
+                span,
+            )),
+        };
+
+        let host_result = match &self.host {
+            Some(h) => Ok(h),
+            None => Err(UrlComponents::generate_shell_error_for_missing_parameter(
+                String::from("host"),
+                span,
+            )),
+        };
+
+        Ok(format!(
+            "{}://{}{}{}{}{}{}",
+            scheme_result?,
+            user_and_pwd,
+            host_result?,
+            self.port
+                .map(|p| format!(":{}", p))
+                .as_ref()
+                .unwrap_or(&String::from("")),
+            self.path.as_ref().unwrap_or(&String::from("")),
+            self.query.as_ref().unwrap_or(&String::from("")),
+            self.fragment.as_ref().unwrap_or(&String::from(""))
```

Comment:
Here there is no need to write `String::from("")` and immediately reference it. `Option` allows you to indirectly `Deref` a `String` to `&str`

```suggestion
                .as_deref()
                .unwrap_or_default(),
            self.path.as_deref().unwrap_or_default(),
            self.query.as_deref().unwrap_or_default(),
            self.fragment.as_deref().unwrap_or_default()
```